Implement osu!-styled Markdown container

gagahpangeran opened the pull request:
- [x] Depends on https://github.com/ppy/osu-framework/pull/4443

## Purpose of this PR

- Change and override generic markdown container from osu! framework to match color and style with markdown in osu!web.
- This can be use for many markdown component in osu!, such as comment, news article, and wiki.

## Summary of this PR

- Change font size and font weight in all existing markdown component.
- Change colour in all markdown existing components based on overlay colour scheme.
- Add new component `OsuMarkdownListItem` to provide bullet and number for ordered and undered list. This component didn't exist in o!f's markdown component.

All of the colour and styling reference based on osu!web and mostly from these two files.
- https://github.com/ppy/osu-web/blob/master/resources/assets/less/bem/osu-md.less
- https://github.com/ppy/osu-web/blob/master/resources/assets/less/base.less

## Known issue in this PR

- Italic font doesn't show. (Looks like osu! doesn't have italic font yet).
- Code block and inline code don't use monospace font yet.
- Inline code doesn't have background colour. (I can't find how to change text background colour).

---

I have [WIP branch for wiki overlay](https://github.com/gagahpangeran/osu/tree/wiki-overlay) that will use this markdown container. I make this PR first, because the changes already quite big and wiki overlay implementation itself is not ready yet.

I could try split up the PR if this is too big to be reviewed.

gagahpangeran committed:
add OsuMarkdownContainer

All of the markdown file in osu-wiki have YAML frontmatter. This YAML is
parsed as common markdown paragraph. So we add `UseYamlFrontMatter()` in
markdown pipeline builder to parse YAML as `YamlFrontMatterBlock`.

gagahpangeran committed:
Add OsuMarkdownLinkText

Color from
https://github.com/ppy/osu-web/blob/d56352aeefc412507c3dab7c16e3e3118421b436/resources/assets/less/functions.less#L159-L165

gagahpangeran committed:
add OsuMarkdownTextFlowContainer

gagahpangeran committed:
override CreateTextFlow in OsuMarkdownContainer

gagahpangeran committed:
add test scene for OsuMarkdownContainer

gagahpangeran committed:
add OsuMarkdownFencedCodeBlock

Reference :
https://github.com/ppy/osu-web/blob/d56352aeefc412507c3dab7c16e3e3118421b436/resources/assets/less/bem/osu-md.less#L41-L45

gagahpangeran committed:
add OsuMarkdownSeparator

Reference
https://github.com/ppy/osu-web/blob/d56352aeefc412507c3dab7c16e3e3118421b436/resources/assets/less/bem/osu-md.less#L19-L21

gagahpangeran committed:
add OsuMarkdownQuoteBlock

Reference:
https://github.com/ppy/osu-web/blob/d56352aeefc412507c3dab7c16e3e3118421b436/resources/assets/less/base.less#L7-L10

gagahpangeran committed:
add OsuMarkdownTableCell

Reference :
https://github.com/ppy/osu-web/blob/d56352aeefc412507c3dab7c16e3e3118421b436/resources/assets/less/bem/osu-md.less#L254-L277

gagahpangeran committed:
add OsuMarkdownTable

gagahpangeran committed:
add test link with inline text markdown

gagahpangeran committed:
add unordered list test

gagahpangeran committed:
add OsuMarkdownListItem

gagahpangeran committed:
use OsuMarkdownListItem for ListItemBlock

gagahpangeran committed:
handle list padding in OsuMarkdownListItem

Reference : https://github.com/ppy/osu-web/blob/5b0e3ac3ffce6b1aff8c9a8794db56603e885ef8/resources/assets/less/bem/osu-md.less#L193-L194

gagahpangeran committed:
change OsuMarkdownListItem to composite drawable

gagahpangeran committed:
add bullet marker in OsuMarkdownListItem

gagahpangeran committed:
add level in OsuMarkdownListItem

gagahpangeran committed:
create list marker based on its level

gagahpangeran committed:
add ordered list test

gagahpangeran committed:
add order in OsuMarkdownListItem for ordered list

gagahpangeran committed:
use order number as marker for ordered list

gagahpangeran committed:
add left padding for ordered list

In osu-md.less, this rule style[1] removes padding left in ordered list.
But in this rule style[2], pseudo element `::before` is used as marker
or counter and has minimal width 30px. So we use this as left padding
size.

[1] https://github.com/ppy/osu-web/blob/5b0e3ac3ffce6b1aff8c9a8794db56603e885ef8/resources/assets/less/bem/osu-md.less#L196-L200
[2] https://github.com/ppy/osu-web/blob/5b0e3ac3ffce6b1aff8c9a8794db56603e885ef8/resources/assets/less/bem/osu-md.less#L210-L219

gagahpangeran committed:
change marker size and position for ordered list

gagahpangeran committed:
add long mixed list test

Copied from https://github.com/ppy/osu-wiki/blob/master/wiki/Tournaments/OWC/2020/en.md#tournament-rules

gagahpangeran committed:
change create text marker to virtual

gagahpangeran committed:
use ListBlock IsOrdered to determine ordered or unordered list

gagahpangeran committed:
add inline code colour

Reference : https://github.com/ppy/osu-web/blob/31579d45aa464a439e943250ca0d5fa7e05a1eb6/resources/assets/less/bem/osu-md.less#L12-L17

gagahpangeran committed:
add hover colour to OsuMarkdownLinkText

gagahpangeran committed:
change default font size to 14

Reference :
- https://github.com/ppy/osu-web/blob/31579d45aa464a439e943250ca0d5fa7e05a1eb6/resources/assets/less/bem/osu-md.less#L9
- https://github.com/ppy/osu-web/blob/31579d45aa464a439e943250ca0d5fa7e05a1eb6/resources/assets/less/variables.less#L161

gagahpangeran committed:
add heading test scene

gagahpangeran committed:
add OsuMarkdownHeading

gagahpangeran committed:
change heading font size

Heading 1 : 30px
- https://github.com/ppy/osu-web/blob/31579d45aa464a439e943250ca0d5fa7e05a1eb6/resources/assets/less/base.less#L12-L16

Heading 2 : 26px
- https://github.com/ppy/osu-web/blob/31579d45aa464a439e943250ca0d5fa7e05a1eb6/resources/assets/less/bem/osu-md.less#L133-L134
- https://github.com/ppy/osu-web/blob/31579d45aa464a439e943250ca0d5fa7e05a1eb6/resources/assets/less/variables.less#L169

Heading 3 : 20px
- https://github.com/ppy/osu-web/blob/31579d45aa464a439e943250ca0d5fa7e05a1eb6/resources/assets/less/bem/osu-md.less#L147-L148
- https://github.com/ppy/osu-web/blob/31579d45aa464a439e943250ca0d5fa7e05a1eb6/resources/assets/less/variables.less#L170
- https://github.com/ppy/osu-web/blob/31579d45aa464a439e943250ca0d5fa7e05a1eb6/resources/assets/less/variables.less#L154

Heading 4 : 18px
- https://github.com/ppy/osu-web/blob/31579d45aa464a439e943250ca0d5fa7e05a1eb6/resources/assets/less/bem/osu-md.less#L160-L161
- https://github.com/ppy/osu-web/blob/31579d45aa464a439e943250ca0d5fa7e05a1eb6/resources/assets/less/variables.less#L171
- https://github.com/ppy/osu-web/blob/31579d45aa464a439e943250ca0d5fa7e05a1eb6/resources/assets/less/variables.less#L153

Heading 5 : 16px
- https://github.com/ppy/osu-web/blob/31579d45aa464a439e943250ca0d5fa7e05a1eb6/resources/assets/less/bem/osu-md.less#L174-L175
- https://github.com/ppy/osu-web/blob/31579d45aa464a439e943250ca0d5fa7e05a1eb6/resources/assets/less/variables.less#L172
- https://github.com/ppy/osu-web/blob/31579d45aa464a439e943250ca0d5fa7e05a1eb6/resources/assets/less/variables.less#L152

Heading 6 : 14px
- https://github.com/ppy/osu-web/blob/31579d45aa464a439e943250ca0d5fa7e05a1eb6/resources/assets/less/bem/osu-md.less#L183-L184
- https://github.com/ppy/osu-web/blob/31579d45aa464a439e943250ca0d5fa7e05a1eb6/resources/assets/less/variables.less#L173
- https://github.com/ppy/osu-web/blob/31579d45aa464a439e943250ca0d5fa7e05a1eb6/resources/assets/less/variables.less#L150

gagahpangeran committed:
add emphases test

gagahpangeran committed:
change font weight for bold text

gagahpangeran committed:
change heading font weight

h1 and h2 : Semi Bold (600)
- https://github.com/ppy/osu-web/blob/376cac43a051b9c85ce95e2c446099be187b3e45/resources/assets/less/bem/osu-md.less#L111
- https://github.com/ppy/osu-web/blob/376cac43a051b9c85ce95e2c446099be187b3e45/resources/assets/less/bem/osu-md.less#L135

The rest of heading : Bold (700)
- https://github.com/ppy/osu-web/blob/376cac43a051b9c85ce95e2c446099be187b3e45/resources/assets/less/bem/osu-md.less#L97

bdach commented on osu.Game/Graphics/Containers/Markdown/OsuMarkdownListItem.cs:
```diff
@@ -0,0 +1,90 @@
+        private readonly int order;
+        private readonly bool isOrdered;
+
+        private const float ordered_left_padding = 30;
+        private const float unordered_left_padding = 20;
```
nit: constants usually go above instance members

bdach commented on osu.Game/Graphics/Containers/Markdown/OsuMarkdownListItem.cs:
```diff
@@ -0,0 +1,90 @@
+            AddInternal(marker);
+        }
+
+        protected virtual string CreateTextMarker()
```
nit: not much creating is actually happening here, `GetMarkerText()` seems like a better name (and if this is to be protected virtual, xmldoc is needed)

gagahpangeran replied:
Yeah, I also just realize maybe it should have `level` and `isOrdered` parameter if the method is protected virtual.

bdach replied:
Definitely. Otherwise all that an inheritor can return is pretty much a constant, at least without reflection.

bdach commented on osu.Game/Graphics/Containers/Markdown/OsuMarkdownQuoteBlock.cs:
```diff
@@ -0,0 +1,45 @@
+            background.Colour = colourProvider.Content2;
+        }
+
+        protected override Drawable CreateBackground()
+        {
+            return background = base.CreateBackground();
+        }
```
should have width 2 ([example in the wild](https://osu.ppy.sh/home/news/2017-09-23-mapping-with-rewards-returns), [relevant css rule](https://github.com/ppy/osu-web/blob/376cac43a051b9c85ce95e2c446099be187b3e45/resources/assets/less/base.less#L7-L10))

bdach commented on osu.Game/Graphics/Containers/Markdown/OsuMarkdownQuoteBlock.cs:
```diff
@@ -0,0 +1,45 @@
+            var textFlow = base.CreateTextFlow();
+            textFlow.Margin = new MarginPadding
+            {
+                Top = 10,
+                Bottom = 10,
+                Left = 20,
+                Right = 20,
```
nit: can simplify by using `Horizontal` and `Vertical`

bdach commented on osu.Game/Graphics/Containers/Markdown/OsuMarkdownTableCell.cs:
```diff
@@ -0,0 +1,52 @@
+                RelativeSizeAxes = Axes.X,
+            };
+
+            // TODO : Change font weight to 700 for heading
```
What's the plan for this? Will this be addressed later? Seems like this could be fixed at this point already via a custom `MarkdownTextFlowContainer` - there should be enough extensibility points to do so.

gagahpangeran replied:
I forget about this.

gagahpangeran replied:
Actually, it can't override `CreateTextFlow()` method to use custom `MarkdownTextFlowContainer` because the method is not virtual.

https://github.com/ppy/osu-framework/blob/635e4199dd3e3b4ae8964f057f482aa94862c7ce/osu.Framework/Graphics/Containers/Markdown/MarkdownTableCell.cs#L79-L84

Any suggestions?

frenzibyte replied:
Open a PR making it virtual then, just like the rest of the components.

gagahpangeran replied:
Yes, I'll do it.

bdach commented:
Aside from the above, missing fonts are kneecapping this implementation a little bit; readability is not too great when everything's in Torus. This is probably about the time to consider adding Inter to resources. And a monospace font, as well, but that's a bigger quandary, since web doesn't actually choose one monospace font, but just uses whichever flavour is available on the system:

```
font-family: Menlo (macOS-specific), Monaco (macOS-specific), Consolas (windows-specific), Courier New (windows-specific), monospace (whatever is left available);
```

gagahpangeran commented on osu.Game/Graphics/Containers/Markdown/OsuMarkdownFencedCodeBlock.cs:
```diff
@@ -0,0 +1,44 @@
+        [BackgroundDependencyLoader]
+        private void load(OverlayColourProvider colourProvider)
+        {
+            // TODO : Change to monospace font to match with osu-web
+            background.Colour = colourProvider.Background6;
+            textFlow.Colour = colourProvider.Light1;
```
At first I also thought about this. But when I try something like this for example.

```csharp
    public class OsuMarkdownFencedCodeBlock : MarkdownFencedCodeBlock
    {

        [Resolved]
        private OverlayColourProvider colourProvider { get; set; }

        public OsuMarkdownFencedCodeBlock(FencedCodeBlock fencedCodeBlock)
            : base(fencedCodeBlock)
        {
        }

        protected override Drawable CreateBackground()
        {
            return new Box
            {
                RelativeSizeAxes = Axes.Both,
                Colour = colourProvider.Background6,
            };
        }

        public override MarkdownTextFlowContainer CreateTextFlow()
        {
            var textFlow = base.CreateTextFlow();
            textFlow.Colour = colourProvider.Light1;
            return textFlow;
        }
    }

```

It just doesn't work and crash.

Am I missing something?

bdach replied:
You're right. This is an unfortunate consequence of how DI works. Because the *base* class's BDL is what is accessing `Create{Background,TextFlow}()`, rather than this one's, this class's `[Resolved]` fields won't have been injected yet at that stage, therefore leading to nullrefs.

This *could* be worked around by using `Dependencies.Get<OverlayColourProvider>()`, but I'm not sure I want to recommend that precedent... This feels like a yet-undiscovered hole in how we do injection, if I'm honest. Maybe a framework change that always injects all `[Resolved]` fields before running any BDLs is needed here? Would need further opinions on this from others.

frenzibyte replied:
I agree all `[Resolved]` members should be injected before any BDL is invoked, doesn't sound to cause any harm.

peppy replied:
This could be fixed by making derived components, rather than weirdly assigning the component to a field.

gagahpangeran replied:
@peppy so basically we add private class as derived component inside this class like this

```csharp
    public class OsuMarkdownFencedCodeBlock : MarkdownFencedCodeBlock
    {
        public OsuMarkdownFencedCodeBlock(FencedCodeBlock fencedCodeBlock)
            : base(fencedCodeBlock)
        {
        }

        protected override Drawable CreateBackground() => new CodeBlockBackground();

        public override MarkdownTextFlowContainer CreateTextFlow() => new CodeBlockTextFlowContainer();

        private class CodeBlockBackground : Box
        {
            [BackgroundDependencyLoader]
            private void load(OverlayColourProvider colourProvider)
            {
                RelativeSizeAxes = Axes.Both;
                Colour = colourProvider.Background6;
            }
        }

        private class CodeBlockTextFlowContainer : OsuMarkdownTextFlowContainer
        {
            [BackgroundDependencyLoader]
            private void load(OverlayColourProvider colourProvider)
            {
                Colour = colourProvider.Light1;
                Margin = new MarginPadding(10);
            }
        }
```

or am I wrong?

smoogipoo replied:
That's correct.

gagahpangeran replied:
The drawback of this pattern is I should re-apply all of customization, for example, from `base.CreateTextFlow()` into the custom derived component `CodeBlockTextFlowContainer`. But after a slightly check in o!f, mostly it just one or two customization.

gagahpangeran committed:
move list item constant position

gagahpangeran committed:
remove unused this

gagahpangeran committed:
rename CreateTextMarker to GetTextMarker

gagahpangeran committed:
add level and isOrdered parameter

gagahpangeran committed:
add xmldoc for GetTextMarker

gagahpangeran committed:
change block quote backgroudn width

Reference : https://github.com/ppy/osu-web/blob/376cac43a051b9c85ce95e2c446099be187b3e45/resources/assets/less/base.less#L7-L10

gagahpangeran committed:
change quote block margin padding to use vertical and horizontal

gagahpangeran committed:
change font weight of table header

gagahpangeran commented on osu.Game/Graphics/Containers/Markdown/OsuMarkdownContainer.cs:
```diff
@@ -0,0 +1,70 @@
+
+namespace osu.Game.Graphics.Containers.Markdown
+{
+    public class OsuMarkdownContainer : MarkdownContainer
```
Actually after checking in osu!web, the margin bottom of the paragraph is 1.5em.

- https://github.com/ppy/osu-web/blob/376cac43a051b9c85ce95e2c446099be187b3e45/resources/assets/less/bem/osu-md.less#L230
- https://github.com/ppy/osu-web/blob/376cac43a051b9c85ce95e2c446099be187b3e45/resources/assets/less/variables.less#L58

Because the base font size 14px, so the actual margin bottom is 1.5 * 14 = 21px.

I crosscheck with devtools and the actual margin bottom is correct 21px.

![Screenshot from 2021-05-06 09-08-06](https://user-images.githubusercontent.com/4964985/117232082-ae4c9700-ae4a-11eb-8eb7-8f0f023a0654.png)

I know line space and margin is a bit different, but I try to compare paragraph in osu!web with line spacing 10 and 21.

| osu!web | spacing 10 | spacing 21 |
| --- | --- | --- |
| ![Screenshot from 2021-05-06 09-16-41](https://user-images.githubusercontent.com/4964985/117232791-1a7bca80-ae4c-11eb-94a6-0792b923c891.png) | ![Screenshot from 2021-05-06 09-15-38](https://user-images.githubusercontent.com/4964985/117232804-236c9c00-ae4c-11eb-912b-ef19660ae428.png) | ![Screenshot from 2021-05-06 09-13-50](https://user-images.githubusercontent.com/4964985/117232813-28c9e680-ae4c-11eb-8d59-f13c57b53f71.png) |

After some trial and error, I get value between 17 or 18.

| osu!web | spacing 18 | spacing 17 |
| --- | --- | --- |
| ![Screenshot from 2021-05-06 09-16-41](https://user-images.githubusercontent.com/4964985/117232791-1a7bca80-ae4c-11eb-94a6-0792b923c891.png) | ![Screenshot from 2021-05-06 09-22-31](https://user-images.githubusercontent.com/4964985/117233069-9c6bf380-ae4c-11eb-99aa-e56d77d0f889.png) | ![Screenshot from 2021-05-06 09-23-39](https://user-images.githubusercontent.com/4964985/117233228-e7860680-ae4c-11eb-9c09-982d2db2d27a.png) |

I don't know if this spacing is different on another screen resolution.

peppy replied:
These numbers should not be trial-and-error, though. The metrics should be transferred directly from osu-web (as they are done in the user profile overlay and other common implementations).

gagahpangeran replied:
Yes, I propose to use 21px value from margin bottom value in osu-web as line spacing here. But as you can see in first table comparison, the spacing is a bit different between **osu-web** and **spacing 21**.

peppy replied:
That's fine. The numbers should match, even if visually it looks different.

peppy commented on osu.Game/Graphics/Containers/Markdown/OsuMarkdownHeading.cs:
```diff
@@ -0,0 +1,76 @@
+
+        protected override float GetFontSizeByLevel(int level)
+        {
+            const float base_font_size = 14;
```
Worth including an inline comment referencing where these sizes come from (ie. a link to the osu-web source)

peppy commented on osu.Game/Graphics/Containers/Markdown/OsuMarkdownFencedCodeBlock.cs:
```diff
@@ -0,0 +1,44 @@
+            };
+        }
+
+        public override MarkdownTextFlowContainer CreateTextFlow()
```
Shouldn't this be using an `OsuMarkdownTextFlowContainer`?

gagahpangeran replied:
`base.CreateTextFlow()` is called `parentFlowComponent.CreateTextFlow()` to create the text flow.

`parentFlowComponent` itself is `OsuMarkdownContainer` and method `CreateTextFlow()` is already using `OsuMarkdownTextFlowContainer` in this PR.

https://github.com/ppy/osu/pull/12688/files/0d3ca8dde197428a2eef2ec2bc4d49d614ccea28#diff-757c2f47d90e82f18a6fa25089a827956e00ec82fdb35a05327d80902b378ce7R45

peppy commented on osu.Game/Graphics/Containers/Markdown/OsuMarkdownLinkText.cs:
```diff
@@ -0,0 +1,48 @@
+
+        public override SpriteText CreateSpriteText()
+        {
+            return spriteText = base.CreateSpriteText();
```
Should be an `OsuSpriteText`

gagahpangeran replied:
Same as https://github.com/ppy/osu/pull/12688/files/0d3ca8dde197428a2eef2ec2bc4d49d614ccea28#r627177370

peppy commented on osu.Game/Graphics/Containers/Markdown/OsuMarkdownListItem.cs:
```diff
@@ -0,0 +1,96 @@
+        {
+            this.level = level;
+            order = listItemBlock.Order;
+            isOrdered = ((ListBlock)listItemBlock.Parent).IsOrdered;
```
This should be split out into two components, rather than conditionaled multiple times in this file.

gagahpangeran replied:
I do this for simplicity to access `IsOrdered` field in `ListBlock`. Fortunately the `Parent` type of `ListItemBlock` is definitely `ListBlock`.

bdach replied:
The cast is not the problem, the fact that bulleted and numbered lists are one component is.

gagahpangeran replied:
Ah yeah, that could be done like that.

Or I can try approach like HTML `<ul>` and `<ol>` as `OsuMarkdown{Ordered,Unordered}List` that have `OsuMarkdownListItem` as its children.

bdach replied:
Sure, I guess. Whichever works best.

peppy commented on osu.Game/Graphics/Containers/Markdown/OsuMarkdownQuoteBlock.cs:
```diff
@@ -0,0 +1,45 @@
+
+        public override MarkdownTextFlowContainer CreateTextFlow()
+        {
+            var textFlow = base.CreateTextFlow();
```
you can use
```csharp
return base.CreateTextFlow().With(f => f.Margin = ...);
```
for cases like this to avoid the cumbersome logic

gagahpangeran committed:
add comment for base font size heading

gagahpangeran committed:
simplify `CreateTextFlow` in quote block

gagahpangeran committed:
add paragraph test

gagahpangeran committed:
change line spacing to 21

We use margin bottom in osu-web markdown paragraph[1] as reference for
this line spacing value.

The value from osu-web itself is 1.5em[2]. Because the base font size of
the paragraph is 14px[3][4], the actual value is 14 * 1.5 = 21px

[1] https://github.com/ppy/osu-web/blob/376cac43a051b9c85ce95e2c446099be187b3e45/resources/assets/less/bem/osu-md.less#L230
[2] https://github.com/ppy/osu-web/blob/376cac43a051b9c85ce95e2c446099be187b3e45/resources/assets/less/variables.less#L58
[3] https://github.com/ppy/osu-web/blob/376cac43a051b9c85ce95e2c446099be187b3e45/resources/assets/less/bem/osu-md.less#L9
[4] https://github.com/ppy/osu-web/blob/376cac43a051b9c85ce95e2c446099be187b3e45/resources/assets/less/variables.less#L161

gagahpangeran committed:
Merge branch 'master' into osu-markdown

gagahpangeran committed:
use derived component in OsuMarkdownFencedCodeBlock

gagahpangeran committed:
add QuoteBackground in OsuMarkdownQuoteBlock

gagahpangeran committed:
add Separator component in OsuMarkdownSeparator

gagahpangeran committed:
change OsuMarkdownListItem to abstract class

gagahpangeran committed:
add OsuMarkdownOrderedListItem

gagahpangeran committed:
add OsuMarkdownUnorderedListItem

gagahpangeran committed:
change CreateListItem method in OsuMarkdownContainer

gagahpangeran committed:
simplify CreateSpriteText in markdown heading

gagahpangeran committed:
simplify CreateSpriteText in markdown table cell

gagahpangeran committed:
simplify CreateEmphasisedSpriteText

gagahpangeran committed:
add return xmldoc in markdown unordered list

gagahpangeran committed:
extract out table head and body border into separate component

gagahpangeran committed:
add CreateBorder method in markdown table cell

smoogipoo reviewed (approved):
I think this is fine to get in as is and work out the kinks as we go. Thanks!

For example: https://github.com/ppy/osu-framework/issues/4447

smoogipoo committed:
Merge branch 'master' into osu-markdown
